lib/tree.rb

```diff
@@ -16,40 +16,105 @@ def initialize
+  # Time Complexity: O(logn) in worst case scenario
+  # Space Complexity: O(1) always just adding 1 new node
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(logn) since binary
+  # Space Complexity: O(1) since adding nothing new
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if key == current_node.key
+
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else
+      find_helper(current_node.right, key)
+    end
+  end
+
+
+  # Time Complexity: O(n) since going to each node once, backtracking but still backtracking a smaller number of times than there are nodes, so it depends on the number of nodes no matter what.
+  # Space Complexity: O(1) - nothing being made
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # helper
+  def inorder_helper(current_node, list)
+    # left, root, right
+    return list if current_node.nil?
+    # recursion! left side first
+    inorder_helper(current_node.left, list)
+    list << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: same as inorder
+  # Space Complexity: same as inorder
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    # root, left, right
+    return list if current_node.nil?
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: same as inorder
+  # Space Complexity: same as inorder
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    # left, right, root
+    return list if current_node.nil?
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+    return list
+  end
+
+  # Time Complexity: O(n) since each node is visited once
+  # Space Complexity: O(1) since it doesn't make anything new
   def height
```

Comment:
The method works, but you have recursion and so some space complexity.  Predictions?